AFQ/bundle_rec/recognize.py

```diff
@@ -0,0 +1,310 @@
+        rng = np.random.RandomState(rng)
+
+    if (save_intermediates is not None) and \
+            (not op.exists(save_intermediates)):
+        os.makedirs(save_intermediates, exist_ok=True)
+
+    logger.info("Preprocessing Streamlines")
+    tg = abu.read_tg(tg, nb_streamlines)
+
+    # If resampling over-write the sft:
+    if nb_points:
+        tg = StatefulTractogram(
+            dps.set_number_of_points(tg.streamlines, nb_points),
+            tg, tg.space)
+
+    if not isinstance(bundle_dict, BundleDict):
+        bundle_dict = BundleDict(bundle_dict)
+
+    tg.to_vox()
+    n_streamlines = len(tg)
+    bundle_decisions = np.zeros(
+        (n_streamlines, len(bundle_dict)),
+        dtype=np.bool8)
+    bundle_to_flip = np.zeros(
+        (n_streamlines, len(bundle_dict)),
+        dtype=np.bool8)
+    bundle_roi_dists = -np.ones(
+        (
+            n_streamlines,
+            len(bundle_dict),
+            bundle_dict.max_includes),
+        dtype=np.uint32)
+
+    fiber_groups = {}
+    meta = {}
+
+    preproc_imap = get_preproc_plan(img, tg, dist_to_waypoint, dist_to_atlas)
+
+    logger.info("Assigning Streamlines to Bundles")
+    for bundle_idx, bundle_name in enumerate(
+            bundle_dict.bundle_names):
+        logger.info(f"Finding Streamlines for {bundle_name}")
+        run_bundle_rec_plan(
+            bundle_dict, tg, mapping, img, reg_template, preproc_imap,
+            bundle_name, bundle_idx, bundle_to_flip, bundle_roi_dists,
+            bundle_decisions,
+            clip_edges=clip_edges,
+            parallel_segmentation=parallel_segmentation,
+            rb_recognize_params=rb_recognize_params,
+            prob_threshold=prob_threshold,
+            refine_reco=refine_reco,
+            rng=rng,
+            return_idx=return_idx,
+            filter_by_endpoints=filter_by_endpoints,
+            save_intermediates=save_intermediates,
+            cleaning_params=cleaning_params)
+
+    if save_intermediates is not None:
+        os.makedirs(save_intermediates, exist_ok=True)
+        bc_path = op.join(save_intermediates,
+                          "sls_bundle_decisions.npy")
+        np.save(bc_path, bundle_decisions)
+
+    conflicts = np.sum(np.sum(bundle_decisions, axis=1) > 1)
+    if conflicts > 0:
+        logger.warning((
+            "Conflicts in bundle assignment detected. "
+            f"{conflicts} conflicts detected in total out of "
+            f"{n_streamlines} total streamlines."
+            "Defaulting to whichever bundle appears first"
+            "in the bundle_dict."))
+    bundle_decisions = np.concatenate((
+        bundle_decisions, np.ones((n_streamlines, 1))), axis=1)
+    bundle_decisions = np.argmax(bundle_decisions, -1)
+
+    # We do another round through, so that we can orient all the
+    # streamlines within a bundle in the same orientation with respect to
+    # the ROIs. This order is ARBITRARY but CONSISTENT (going from ROI0
+    # to ROI1).
```

Comment:
This is no longer ARBITRARY! And in general, this is a lot more systematic as of recently.

Reply:
Yes, and we have documented the LPI convention elsewhere. If the user provides their own bundle, however, it could be in any orientation. Also, we don't necessarily use ROIs, although we can. So I think I will remove this comment and instead give a more general one.